### PR TITLE
ZEP1: Apply review feedback for terminology section

### PR DESCRIPTION
**Pull request by jstriebel**
This PR applies some feedback from the ZEP1 review in https://github.com/zarr-developers/zarr-specs/pull/149. It adds illustrations to the terminology, and improves the store and storage transformer sections. In detail:

* Added two illustrations, one for the general hierarchy and array terminology, and one for the codec/array/storage transformers/store workflow. Both were made with https://excalidraw.com. The png files include the source and can be loaded into excalidraw for further editing.
* Added examples to store terminology section, as requested in
  https://github.com/zarr-developers/zarr-specs/pull/149#discussion_r928313883
* Improved the storage transformer terminology section, as requested in
  https://github.com/zarr-developers/zarr-specs/pull/149#discussion_r928314024
* Reordered the terminology definitions slightly, to fit better to the illustrations
  (moved codec and compressor more to the end, after the second illustration).

----------------

## Screenshots of the rendered specs with the illustrations
(The changes from the ZEP1 review were applied to the main branch for the screenshots)

Screenshot from 2022-10-04 09-43-20

--------------

Screenshot from 2022-10-04 09-43-43

### CODE REVIEW

**Comment by jstriebel** [comment]
This PR replaces the original PR https://github.com/alimanfoo/zarr-specs/pull/1, as we discussed to gather all PRs here onto the `main` branch (see https://github.com/zarr-developers/zarr-specs/pull/149#issuecomment-1302440391).

**Review by joshmoore — APPROVED** [review]
* "Added two illustrations": :+1: and in general :100: for any visual material since it helps the reader. (Downside will be keeping it in sync long-term)
  - "The png files include the source" Interesting!
* Added examples to store terminology section :+1:
* Improved the storage transformer terminology section :+1:
* Reordered the terminology definitions slightly :+1:

There's still a large re-reading outstanding from my side, but I don't see a reason not to get these in.

File: docs/core/v3.0.rst [hunk 1]
```diff
@@ -360,6 +367,44 @@ conceptual model underpinning the Zarr format.
...
+*Metadata document*
+
+    Each array_ in a hierarchy_ is represented by a metadata document,
+    which is a machine-readable document containing essential
+    processing information about the node. For example, an array_
+    metadata document will specify the number of dimensions_, shape_,
+    `data type`_, grid_, `memory layout`_ and compressor_ for that
+    array_.
+
+    Groups can have a optional metadata document which provide extra
```
Comment (rabernat) [inline docs/core/v3.0.rst:382]:
```suggestion
    Groups can have an optional metadata document which provides extra
```

File: docs/core/v3.0.rst [hunk 2]
```diff
@@ -360,6 +367,44 @@ conceptual model underpinning the Zarr format.
...
+    information about a group.
+
+.. _store:
+.. _stores:
+
+*Store*
+
+    The `metadata documents`_ and encoded chunk_ data for all nodes in a
+    hierarchy_ are held in a store as a sequence of bytes. To enable a variety
```
Comment (rabernat) [inline docs/core/v3.0.rst:391]:
```suggestion
    hierarchy_ are held in a store as raw bytes. To enable a variety
```

File: docs/core/v3.0.rst [hunk 3]
```diff
@@ -360,6 +367,44 @@ conceptual model underpinning the Zarr format.
...
+    of different store types to be used, this specification defines an
+    `Abstract store interface`_ which is a common set of operations that stores
+    may provide. For example, a directory in a file system can be a zarr store,
+    where keys are file names, values are file contents, and files can be read,
+    written, listed or deleted via the operating system. Equally, an S3 bucket
+    can provide this interface, where keys are resource names, values are
+    resource contents, and resources can be read, written or deleted via HTTP.
+
+The following figure illustrates the codec, store and storage transformer
+terminology on the example of reading from an array:
```
Comment (rabernat) [inline docs/core/v3.0.rst:401]:
```suggestion
terminology for a use case of reading from an array:
```

File: docs/core/v3.0.rst [hunk 4]
```diff
@@ -384,43 +429,20 @@ conceptual model underpinning the Zarr format.
...
-    See the `storage transformers details`_ below.
+    To enhance the storage capabilities, storage transformers may change the
+    storage structure and behaviour of data coming from an array_ in the
+    underlying store_, being able to manipulate the stored bytes and the key
+    they are stored at. Upon retrieval the original data is restored within the
+    transformer. Any number of `predefined storage transformers`_ can be
+    registered and stacked. They allow to have common behavior that is not store
+    specific, but in the contrast to codecs they rather affect the storage
+    format of a complete array, rather than single chunks. See the
+    `storage transformers details`_ below.
```
Comment (rabernat) [inline docs/core/v3.0.rst:445]: I find this a bit hard to parse. Suggested revision:

```suggestion
    To provide performance enhancements or other optimizations,
    storage transformers may intercept and alter the storage keys and bytes
    of an array_ before they reach the underlying physical storage.
    Upon retrieval, the original keys and bytes are restored within the
    transformer. Any number of `predefined storage transformers`_ can be
    registered and stacked. In contrast to codecs, stroage transformers act on
    the a complete array, rather than individual chunks. See the
    `storage transformers details`_ below.
```

**Review by rabernat — APPROVED** [review]
I think this is excellent, particularly the drawings. Thanks so much @jstriebel for these additions, which make the spec much more clear.

I have some minor editorial suggestions.

**Review by MSanKeys963 — DISMISSED** [review]

**Comment by MSanKeys963** [comment]
Thanks for sending this PR, @jstriebel.
The illustrations are excellent, and the text looks clear to me. I have one minor feedback:

If I try to view the illustrations [here](https://github.com/scalableminds/zarr-specs/blob/ZEP0001-apply-storage-transformer-feedback-main/docs/core/v3.0.rst), I can't view them properly. This is because they don't have a background colour.
Should we add the illustrations with white background?

PS. I don't have a strong opinion on this, as I know these illustrations can be viewed on the web browser just fine. It's just the `.md` file.

File: docs/core/v3.0.rst [hunk 5]
```diff
@@ -434,14 +434,13 @@ terminology on the example of reading from an array:
...
-    they are stored at. Upon retrieval the original data is restored within the
+    To provide performance enhancements or other optimizations,
+    storage transformers may intercept and alter the storage keys and bytes
+    of an array_ before they reach the underlying physical storage.
+    Upon retrieval, the original keys and bytes are restored within the
     transformer. Any number of `predefined storage transformers`_ can be
-    registered and stacked. They allow to have common behavior that is not store
-    specific, but in the contrast to codecs they rather affect the storage
-    format of a complete array, rather than single chunks. See the
+    registered and stacked. In contrast to codecs, stroage transformers act on
```
Comment (rabernat) [inline docs/core/v3.0.rst:442]:
```suggestion
    registered and stacked. In contrast to codecs, storage transformers act on
```

I think this is my typo

Comment (jstriebel) [reply docs/core/v3.0.rst]: Fixed it already, thanks :-)

**Comment by jstriebel** [comment]
> If I try to view the illustrations [here](https://github.com/scalableminds/zarr-specs/blob/ZEP0001-apply-storage-transformer-feedback-main/docs/core/v3.0.rst), I can't view them properly. This is because they don't have a background colour.
Should we add the illustrations with white background?

@MSanKeys963 Good point, next time I'll add the white background, but I'd leave it as-is for now for simplicity.

**Review by joshmoore — APPROVED** [review]
